### PR TITLE
Cleaned-up log4j 1.2 that disables scary networking (base=1.2.17, fully binary compatible)

### PR DESCRIPTION
**Pull request by lsimons**
Conservative version of earlier PR #16. This has full binary and source compatibility and starts of from 1.2.17, not including the non-released changes made many years ago after the release of 1.2.17.

Unlike PR16, this branch is neatly structured as a series of minimal mostly-idempotent patches, that could be accepted step by step:

*cleanup-only patches*
1. [https://github.com/lsimons/log4j/tree/feat/2021-modern-build](2021-modern-build): make build work with maven 3
2. [https://github.com/lsimons/log4j/tree/fix/2021-robust-tests](2021-robust-tests): clean up the test suite
3. [https://github.com/lsimons/log4j/tree/feat/2021-update-website](2021-update-website): clean up the website
4. [https://github.com/lsimons/log4j/tree/fix/2021-chainsaw-serialVersionUID](2021-chainsaw-serialVersionUID): binary compatibility for chainsaw package. This change is not contained in PR16, since chainsaw was removed on trunk after the release of 1.2.17.

*security fixes*
1. [https://github.com/lsimons/log4j/tree/fix/2021-secure-net](2021-securenet): conservative source/binary compatible versions of the security fixes to log4j.net from PR16.
2. [https://github.com/lsimons/log4j/tree/fix/2021-secure-jmx](2021-secure-jmx): conservative source/binary compatible versions of the security fix to log4j.jmx from PR16.
3. [https://github.com/lsimons/log4j/tree/fix/2021-secure-jdbc](2021-secure-jdbc): security fix to log4j.jdbc from PR16
4. [https://github.com/lsimons/log4j/tree/fix/2021-secure-erfa](2021-secure-erfa): security fix to ExternallyRolledFileAppender from PR16.

*bugfixes*
1. [https://github.com/lsimons/log4j/tree/fix/2021-mdc-java-version](2021-mdc-java-version): bugfix to Loader.java/MDC.java to allow it to work properly on JDK >= 9 from PR16 .

Please note that while this PR is filed against trunk, it can't be merged directly because trunk has moved on from 1.2.17. Suggested approach is to make a bugfix 1.2.x branch starting from tag v1_2_17, and then apply these changes to it, and release 1.2.18 from there.

### CODE REVIEW

**Comment by garydgregory** [comment]
-1 
This is completely unacceptable. You cannot just throw away major blocks of functionality.
I'll reply to the thread message on the mailing list.

File: src/main/java/org/apache/log4j/jdbc/JDBCAppender.java [hunk 1]
```diff
@@ -17,382 +17,127 @@
...
-     sql option value.
-
-    </ul>
-
-    @author Kevin Steppe (<A HREF="mailto:ksteppe@pacbell.net">ksteppe@pacbell.net</A>)
+  Changed in 1.2.18+ to complain about its use and do nothing else.
+  See <a href="https://logging.apache.org/log4j/1.2/">the log4j 1.2 homepage</a>
+  for more information on why this class is disabled since 1.2.18.
 
+  @author Kevin Steppe (<A HREF="mailto:ksteppe@pacbell.net">ksteppe@pacbell.net</A>)
```
Comment (garydgregory) [inline src/main/java/org/apache/log4j/jdbc/JDBCAppender.java:35]: We no longer use author tags in general. Credit is attributed in pom.xml and/or changes.xml.

Comment (lsimons) [reply src/main/java/org/apache/log4j/jdbc/JDBCAppender.java:35]: I didn't add any author tags, but also didn't remove any. Actual change here is only consistent whitespace indenting, Github diff is not displaying it too cleanly.

File: src/main/java/org/apache/log4j/net/SMTPAppender.java [hunk 2]
```diff
@@ -154,7 +160,29 @@ void activateOptions() {
...
+      if ("smtps".equals(smtpProtocol)) {
+        LogLog.warn("WARN-LOG4J-NETWORKING-REMOTE-SMTPS: logging to remote SMTPS host '" +
+            smtpHost + "'! Log4J 1.2 cannot verify the host, upgrade to Log4J 2!");
+      } else {
+        LogLog.warn("WARN-LOG4J-NETWORKING-REMOTE-SMTP: logging to remote SMTP host '" +
+            smtpHost + "'! SMTP is an unencrypted protocol.");
+      }
+    }
+  }
+
```
Comment (Sanix-Darker) [inline src/main/java/org/apache/log4j/net/SMTPAppender.java:185]: 👍🏾

**Comment by larrywest** [comment]
_My $0.000002:_

Log4j 1.x reached end-of-life long ago - last updated a decade ago.

> On August 5, 2015 the Logging Services Project Management Committee announced that Log4j 1.x had reached end of life

From 2019 in re CVE-2019-17571:

> ... This can provide an attack vector that can be expoited. Since Log4j 1 is no longer maintained this issue will not be fixed. Users are urged to upgrade to Log4j 2.

* https://logging.apache.org/log4j/1.2/
* https://blogs.apache.org/foundation/entry/apache_logging_services_project_announces

There's no rationale given here for bringing log4j 1.x back from the dead, and anyone who cares about security or project hygiene has long since moved on to log4j2 or logback.

Assuming there are reasons, why not fork this and call it something else?

**Comment by lsimons** [comment]
The rationale is not found here because most discussion about what should happen with apache projects happens on mailing lists. In this case, the discussion can be followed (and joined!) on dev@logging.apache.org, archives at https://lists.apache.org/list.html?dev@logging.apache.org , more info at https://logging.apache.org/log4j/2.x/mail-lists.html .

**Comment by lsimons** [comment]
Superceded by #18

**Comment by lsimons** [comment]
See
* https://reload4j.qos.ch/
* https://github.com/qos-ch/reload4j

for a maintained/released fork with similar security fixes.

**Pull request closed without merging**